pkg/registry/apis/datasource/utils.go

```diff
@@ -0,0 +1,22 @@
+package datasource
+
+import (
+	"errors"
+	"runtime/debug"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+func panicGuard[R *backend.CheckHealthResult | *backend.QueryDataResponse | interface{}](logger log.Logger, f func() (R, error)) (res R, err error) {
```

Comment:
nice!

rather than making this datasource specific... what about a panic guard for`http.HandlerFunc` -- that way we can catch all these in a consistent way.

eg:
https://github.com/grafana/grafana/blob/main/pkg/registry/apis/query/query.go#L88
and
https://github.com/grafana/grafana/blob/main/pkg/registry/apis/datasource/sub_query.go#L56

seems appropriate to add a "panicGuard" to:
https://github.com/grafana/grafana/tree/main/pkg/apimachinery/utils

Reply:
I discussed similar thing with Andres in DM

> Seems though like this is as planned in k8s apiserver so maybe good to discuss with @toddtreece  if we want to change it for any api group/endpoint https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go#L31

Adding a filter/middleware here https://github.com/grafana/grafana/blob/3532d13f0cf2fcfa29c75416e75f25b78d1b6968/pkg/apiserver/builder/helper.go#L84 might also be an alternative

Reply:
Turns out that we are already using the filter [`WithPanicRecovery`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go#L31) as part of the [`DefaultBuildHandlerChain`](https://github.com/grafana/grafana/blob/apisDatasourcePanic/pkg/apiserver/builder/helper.go#L98C30-L98C54). The problem is that recovery [still panics at the end](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/runtime/runtime.go#L87). If we set that `ReallyCrash` to false we avoid it and the client gets the default response (and the panic gets logged):

```
$ curl -k https://localhost:6445/apis/testdata.datasource.grafana.app/v0alpha1/namespaces/stack-101/connections/ador449jkgohse/health
This request caused apiserver to panic. Look in the logs for details.
```

Which is not a JSON response but it's better than simply closing the connection. We could write our own panic recovery logic but we'd miss what the [k8s implementation does by default](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go#L32).

I'm going to update this PR to simply not crash, let me know what you think.

Reply:
Nice find 👍